aiida/tools/visualization/graph.py

```diff
@@ -495,20 +533,38 @@ def add_incoming(self, node, link_types=(), annotate_links=None, return_pks=True
         :returns: list of nodes or node pks
 
         """
+        # pylint: disable=too-many-locals
         if annotate_links not in [None, False, 'label', 'type', 'both']:
             raise AssertionError('annotate_links must be one of False, "label", "type" or "both"')
 
-        node = self.add_node(node)
+        graph_traversal_rules = self._link_types_to_traversal_rules(link_types=link_types, direction='backward')
+        node_pk = node if isinstance(node, int) else node.pk
+
+        traversed_graph = traverse_graph((node_pk,), max_iterations=1, get_links=True, **graph_traversal_rules)
 
-        nodes = []
-        for link_triple in node.get_incoming(link_type=self._convert_link_types(link_types)).link_triples:
-            self.add_node(link_triple.node)
-            link_pair = LinkPair(link_triple.link_type, link_triple.link_label)
-            style = self._link_styles(
+        traversed_nodes = QueryBuilder().append(
+            Node, filters={'id': {
+                'in': traversed_graph['nodes']
+            }}, project=['id', '*'], tag='node'
+        )
+        traversed_nodes = {query_result[0]: query_result[1] for query_result in traversed_nodes.all()}
```

Comment:
Paging @zooks97

Reply:
This is a workaround for #3679, and is actually the reason I found the specific case for the problem in #3679. If you fix that issue, we should definitely use `traversed_nodes.dict()`.

Reply:
So, even after the fix, I think the `dict()` method is still not very suitable for what we want to do. This is because the dict returned doesn't have the structure we require:

`{ id: instance, ... }`

but instead has a list of more complicated dicts:

`[ {'node': {'id': id, '*': instance} }, ... ]`.

I think the parsing of that would be even more annoying than the current parsing of the method `all()`, which is easier to transform into what we need:

 `[ [id, instance], ... ]`.